### ISSUE TITLE
Joint limit awareness by weighting

### ISSUE THREAD

**Issue by polarursus** [bourumir-wyngs/rs-opw-kinematics#1]
Hello,

This is a nice port of the original OPW library, thank you for the publication. 
I'm using the orignial C++ version for a while now, I wanted to test out your sorting solution.

Do you have an idea how to be aware of joint limits while choosing closest solution?

Here is a video where I'm testing your library with linear interpolation on a straight line that has two wrist singularity points. You can see that on the second point we are saturating J4 hitting negative limit of -270 degrees.
https://www.youtube.com/watch?v=Mt3tKbL4G4I

The IK result goes beyond this, so it gives a valid solution for the Pose, my limiter stops the axis movement.

Would it be a possible solution to have some kind of weighing while sorting, that favors solutions that are going away from joint limits when it is close (calculating direction using previous data and calculate weights by how close the joint to the limit)? 

Many thanks,

**Comment by bourumir-wyngs** [comment]
I am currently working on constraints. There is a working (tests pass) code in a branch 1.1.0 that is not yet released. However, so far it is only able to discard solutions that are outside the specified boundaries (not many joints could rotate 360 degrees). You are asking for the more complex thing. I will think how to implement this. Looks like a good idea.

**Comment by polarursus** [comment]
Well, yes, discarding solutions beyond limits is a great step to keep things safe, but this still could cause a real robot to stop in these constellations, needing to unwind J4 and J6 manually to continue. 

I believe that it would be also awesome to be able to restrict the solutions to one certain axis configuration to keep the arm in a predictable shape. Now for example if you start from 0,0,0,0,0,0 position, the arm could go concave or convex with J2 and J3 depending on the angle of the TCP movement line, I believe this should be restricted as it can cause unwanted movements / joint configurations in certain scenarios and hit something in the workspace.

So use only one solution that the IK solver generates or maybe add the other wrist result to the possible solutions (using the other result that has the same values for J1, J2, J3).

The other valid solutions have to be sorted could be the joint configurations where the FK gives the same result due to longer axis spans. So generate all the feasible solutions, for example -110° on J4 would give the same result like it would be 250°, and so.

I hope that this input helps you rather than messing up your mind :)

Let me know if I can test something with my setup.

Keep up the good work!

**Comment by bourumir-wyngs** [comment]
The recently released version 1.1.0 supports constraints (joint limits) and allows the weighted rating by both proximity to constraint center and the previous position. The constants _BY_PREV_ ( = 0.0) and _BY_CONSTRAINTS_ ( = 1.0) are for convenience only. Intermediate values like 0.6 can also be specified and result in weighted sorting:

```Rust
  let robot = OPWKinematics::new_with_constraints(
    Parameters::irb2400_10(), Constraints::new(
      [-0.1, 0.0, 0.0, 0.0, -PI, -PI],
      [ PI, PI, 2.0*PI, PI, PI, PI],
      BY_CONSTRAINS,
    ));
```
Thanks for pointing out to the need of this important feature.
